arch/rv32i/src/lib.rs

```diff
@@ -81,25 +84,24 @@ pub unsafe fn init_memory() {
     tock_rt0::zero_bss(&mut _szero, &mut _ezero);
 }
 
-/// Tell the MCU what address the trap handler is located at. The trap handler
-/// is called on exceptions and for interrupts.
+pub enum PermissionMode {
+    Machine = 0x1,
+    Supervisor = 0x10,
+    User = 0x100,
+}
+
+/// Tell the MCU what address the trap handler is located at.
 ///
-/// This is a generic implementation. There may be board specific versions as
-/// some platforms have added more bits to the `mtvec` register.
-pub unsafe fn configure_trap_handler() {
-    asm!("
-        // The csrw instruction writes a Control and Status Register (CSR)
-        // with a new value.
-        //
-        // CSR 0x305 (mtvec, 'Machine trap-handler base address.') sets the
-        // address of the trap handler. We do not care about its old value, so
-        // we don't bother reading it.
-        csrw 0x305, $0        // Write the mtvec CSR.
-    "
-    :
-    : "r"(&_start_trap)
-    :
-    : "volatile");
+/// The trap handler is called on exceptions and for interrupts.
+pub unsafe fn configure_trap_handler(mode: PermissionMode) {
```

Comment:
It would be nice to keep the comment about how this is a generic implementation that specific boards may not want to use.